<?xml version="1.0"?>
<launch>
  <arg name="namespace" default="/"/>
  <arg name="cmd_timeout" default="0.5"/>
  <arg name="port" default="/dev/rosimu" />

<!-- /speed_params is read by ROS2portXfer.py and sent to tractor microcontroller via serial -->
<!--    transmissionFullReversePos,
        transmission075ReversePos,
        transmissionNeutralPos,
        transmission025ForwardPos,
        transmission050ForwardPos,
        transmission075ForwardPos,
        transmissionFullForwardPos,
-->

  <param name="/speed_params" type="yaml" value="[245, 258, 266, 312, 313, 314, 325]" /> 

  <!-- RTK base station coordinates polled in odom_from_wheel_and_gps.py every 60 sec in 'calibrate_lat_lon_origin'-->
  <param name="base_station_lat" value="40.485509842" />
  <param name="base_station_lon" value="-80.332308247" />

  <include file="$(find ackermann_vehicle)/launch/ackermann_xacro.launch">
    <arg name="namespace" value="$(arg namespace)"/>
  </include>

  <group ns="$(arg namespace)">
    <node name="controller_spawner" pkg="controller_manager" type="spawner"
          args="$(find ackermann_vehicle)/config/cub_cadet_spawner_params.yaml"/>
    <node name="ackermann_controller" pkg="ackermann_vehicle" type="ackermann_controller.py">
      <param name="cmd_timeout" value="$(arg cmd_timeout)"/>
      <rosparam file="$(find ackermann_vehicle)/config/cub_cadet_cntrlpy_params.yaml" command="load"/>
    </node>
  </group>

  <node pkg="rosserial_python" type="serial_node.py" name="odom_left_serial_node" output="screen">
    <param name="port" value="/dev/odom_left"/>
    <param name="baud" value="115200"/>
  </node>

  <node pkg="rosserial_python" type="serial_node.py" name="odom_right_serial_node" output="screen">
    <param name="port" value="/dev/odom_right"/>
    <param name="baud" value="115200"/>
  </node>

  <node pkg="chip_imu_driver" type="chip_imu_driver_node" name="chip_imu_driver_node" output="screen">
      <param name="port" value="/dev/rosimu" />
      <param name="initial_yaw" type="double" value="-90.94"/> 
  </node>

  <node name="map_server" pkg="map_server" type="map_server"
       args="$(find ackermann_vehicle)/maps/collins_dr_62.yaml" output="screen">
    <param name="frame_id" value="map"/>
  </node>

  <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_odom" args="0 0 0 0 0 0 map odom" />

  <node name="nmea_serial_driver_node" pkg="nmea_navsat_driver" type="nmea_serial_driver" output="screen">
    <param name="port" value="/dev/gps"/>
    <param name="baud" value="115200" />
    <param name="frame_id" value="gps" />
    <param name="use_GNSS_time" value="False" />
    <param name="time_ref_source" value="gps" />
    <param name="useRMC" value="False" />
  </node>

  <node name="odom_wheel_node" pkg="ackermann_vehicle" type="odom_from_wheel_and_gps.py" output="screen" />

  <node name="serial_xfer" pkg="ackermann_vehicle" type="ROS2portXfer.py" output="log" />  

  <node pkg="pure_pursuit" type="pure_pursuit" name="pure_pursuit_steering_controller"  clear_params="true">
    <param name="robot_frame_id" value="base_link"/>
    <param name="global_frame_id" value="odom"/>
    <param name="lookahead_frame_id" value="lookahead"/>
    <rosparam file="$(find pure_pursuit)/cfg/pure_pursuit_cub_cadet.yaml" command="load"/>    
    <remap from="path_segment" to="/drive_path"/>
    <remap from="odometry" to="odom"/>
    <remap from="cmd_vel" to="cmd_vel"/>
    <remap from="cmd_acker" to="/acker_in"/>
  </node>  

</launch>